File: src/dotnet/Kurdi.ECommerce.Inventory.Core/Entities/ProductAggregate/ProductQuantity.cs
using System.ComponentModel.DataAnnotations.Schema;
using Kurdi.ECommerce.Inventory.Core.Exceptions;
using Microsoft.EntityFrameworkCore;

namespace Kurdi.ECommerce.Inventory.Core.Entities.ProductAggregate
{
    [Owned]
    public class ProductQuantity
    {
        [Column(name:"total_stock")]
        public int TotalStock { get; private set; }

        [Column(name:"available_stock")]
        public int AvailableStock { get; private set; }

        [Column(name:"reserved_stock")]
        public int ReservedStock { get; private set; }

        public void AddStock(int quantity)
        {
            this.TotalStock += quantity;
            this.AvailableStock += quantity;
        }    

        public void ReserveStock(int quantity)
        {
            if(this.AvailableStock - quantity < 0)
            {
                throw new NegativeStockTransactionException();
            }
            this.ReservedStock += quantity;
            this.AvailableStock -= quantity;
        }

        public void CancelReservation(int quantity)
        {
            if (this.ReservedStock - quantity < 0)
            {
                throw new NegativeStockTransactionException();
            }
            this.ReservedStock -= quantity;
            this.AvailableStock += quantity;
        }
    }
}

File: src/main/java/com/kurdi/inventory/domain/entities/products/ProductQuantity.java
package com.kurdi.inventory.domain.entities.products;

import lombok.*;
import javax.persistence.Embeddable;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

@Embeddable
@Data
@AllArgsConstructor
@NoArgsConstructor
public class ProductQuantity implements Serializable {
    private List<StockItem> stock = new ArrayList<>();

    @Setter(AccessLevel.NONE)
    int totalStock = 0;
    @Setter(AccessLevel.NONE)
    int availableStock = 0;
    @Setter(AccessLevel.NONE)
    int reservedStock = 0;

    
    public void addStock(List<StockItem> stock) {

        //add to stock list
        this.stock.addAll(stock);
        //generate sku for all of them
        totalStock += stock.size();
        availableStock += stock.size();
    }

    public void reserveStock(int quantity) {
        availableStock -= quantity;
        reservedStock += quantity;
    }

    public void cancelReservation(int quantity) {
        availableStock += quantity;
        reservedStock -= quantity;
    }

    public void selling(int quantity) {
        availableStock -= quantity;
        totalStock -= quantity;
    }
}
